test/e2e/framework/profile_gatherer.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package framework
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+)
+
+const (
+	// Default value for how long the CPU profile is gathered for.
+	DefaultCPUProfileSeconds = 10
+)
+
+func GetProfilesDirectoryPath() string {
+	return path.Join(TestContext.ReportDir, "profiles")
+}
+
+func CreateProfilesDirectoryIfNeeded() error {
+	profileDirPath := GetProfilesDirectoryPath()
+	if _, err := os.Stat(profileDirPath); os.IsNotExist(err) {
+		if mkdirErr := os.Mkdir(profileDirPath, 0777); mkdirErr != nil {
+			return fmt.Errorf("Failed to create profiles dir: %v", mkdirErr)
+		}
+	} else if err != nil {
+		return fmt.Errorf("Failed to check existence of profiles dir: %v", err)
+	}
+	return nil
+}
+
+func CheckProfileGatheringPrerequisites() error {
+	if !TestContext.AllowGatheringProfiles {
+		return fmt.Errorf("Can't gather profiles as --allow-gathering-profiles is false")
+	}
+	if TestContext.ReportDir == "" {
+		return fmt.Errorf("Can't gather profiles as --report-dir is empty")
+	}
+	if err := CreateProfilesDirectoryIfNeeded(); err != nil {
+		return fmt.Errorf("Failed to ensure profiles dir: %v", err)
+	}
+	return nil
+}
+
+func GatherApiserverCPUProfile(profileBaseName string) error {
+	return GatherApiserverCPUProfileForNSeconds(profileBaseName, DefaultCPUProfileSeconds)
```

Comment:
Default is 30 usually. Do you have any motivation for making default 10?

Reply:
Just a feeling that 30s might be too long than what we may need normally.
Also keeping the SSH connection open for 30s may not be that great an idea. Just my 2 cents.

Reply:
Let's not make our default different from everyone else's default.
Also, why do you think keeping an SSH session open for 30 seconds is too long? It's going to be idle while waiting for data anyway, no?

Reply:
Agreed... changed it to 30. Yeah, it's going to be idle in the meanwhile.